### PR TITLE
XRI3 migration step04

### PR DESCRIPTION
**Pull request by ms-RistoRK**
Aiming for the goal of removing the deprecated ControllerLookup class for XRI3 migration completion in this PR:
* Marked ControllerLookup as obsolete.
* Added TrackedPoseDriverLookup that will be the XRI-3 compliant replacement of ControllerLookup.
* Added a TrackedPoseDriverLookup field for FlatSceenModeDetector and Solver classes.
* Updated HandConstraintPalmUp and FlatSceenModeDetector to handle the presence of the new TrackedPoseDriverLookup and still be backward compatible to ControllerLookup if such component exists. The logic, imho, will help existing MRTK3 clients to migrate their code to XRI3 by doing:
  * If an existing ControllerLookup exists then it has priority to be consumed despite having a TrackedPoseDriverLookup present or not.
  * If a TrackedPoseDriverLookup is present it will be consumed iff a ControllerLookup is not present.
  * Both ControllerLookup and TrackedPoseDriverLookup can be present but ControllerLookup will consumed if both are present.
  I would suggest that the steps for a client to migrate their pre-XRI3 code to XRI3-compliant MRTK3 code be: Add a TrackedPoseDriverLookup to their prefabs -> link interactors TrackedPoseDrivers to TrackedPoseDriverLookup -> disable (do not remove) ControllerLookup -> test -> if everything works as expected then remove ControllerLookup.
* Replaced ControllerLookup component in Experimental-XRI3/MRTKInteractionManager prefab with a TrackedPoseDriverLookup component as shown next:
image

### CODE REVIEW

File: org.mixedrealitytoolkit.spatialmanipulation/Solvers/HandConstraintPalmUp.cs [hunk 1]
```diff
@@ -274,21 +274,47 @@ private bool IsUserGazeMeetingThresholdRequirements(XRNode hand)
                 Ray? gazeRay = null;
                 bool usedEyeGaze = false;
 
-                if (ControllerLookup != null &&
-                    ControllerLookup.GazeController != null &&
-                    (ControllerLookup.GazeController.currentControllerState.inputTrackingState &
-                    (InputTrackingState.Position | InputTrackingState.Rotation)) > 0)
+                if (ControllerLookup != null) //Note: Will be removed for XRI3 migration completion
```
Comment (AMollis) [inline org.mixedrealitytoolkit.spatialmanipulation/Solvers/HandConstraintPalmUp.cs:277]: I think you were right earlier :)  we need branching for the old rig to work.   I think this is a case where you can't remove this.  Surround with `#pragma warning disable`

```cs
#pragma warning disable CS0618 // Type or member is obsolete
if (ControllerLookup != null)
{
    // your code
}
#pragma warning restore CS0618
```

File: org.mixedrealitytoolkit.spatialmanipulation/Solvers/Solver.cs [hunk 2]
```diff
@@ -15,12 +17,21 @@ namespace MixedReality.Toolkit.SpatialManipulation
     [HelpURL("https://docs.microsoft.com/windows/mixed-reality/mrtk-unity/features/ux-building-blocks/solvers/solver")]
     public abstract class Solver : MonoBehaviour
     {
-        private static ControllerLookup controllerLookup;
+        [Obsolete]
```
Comment (AMollis) [inline org.mixedrealitytoolkit.spatialmanipulation/Solvers/Solver.cs:20]: Add message parameter

File: org.mixedrealitytoolkit.spatialmanipulation/Solvers/Solver.cs [hunk 3]
```diff
@@ -15,12 +17,21 @@ namespace MixedReality.Toolkit.SpatialManipulation
...
+        private static ControllerLookup controllerLookup; //Note: Will be removed for XRI3 migration completion
 
         /// <summary>
         /// Get the <see cref="Toolkit.ControllerLookup">ControllerLookup</see> that will be used all application <see cref="Solver"/> objects.
         /// </summary>
-        protected static ControllerLookup ControllerLookup => controllerLookup;
+        [Obsolete]
```
Comment (AMollis) [inline org.mixedrealitytoolkit.spatialmanipulation/Solvers/Solver.cs:26]: Add message parameter

File: org.mixedrealitytoolkit.spatialmanipulation/Solvers/Solver.cs [hunk 4]
```diff
@@ -218,6 +229,11 @@ protected virtual void Awake()
             }
 
             GoalScale = maintainScaleOnInitialization ? transform.localScale : Vector3.one;
+
+            if (trackedPoseDriverLookup == null)
```
Comment (AMollis) [inline org.mixedrealitytoolkit.spatialmanipulation/Solvers/Solver.cs:233]: Why we doing this in `Awake` and `Start`?   I suggest follow the existing pattern used by `controllerLookup`, and only do this in `Start`

File: org.mixedrealitytoolkit.spatialmanipulation/Solvers/Solver.cs [hunk 5]
```diff
@@ -265,6 +281,13 @@ protected virtual void Start()
             {
```
Comment (AMollis) [inline org.mixedrealitytoolkit.spatialmanipulation/Solvers/Solver.cs:281]: I think you were right earlier :)  we need branching for the old rig to work.   I think this is a case where you can't remove this.  Surround with `#pragma warning disable`  .  

Also move the `controllerLookup` assignment after the `trackedPoseDriverLookup` assignment, and only if `trackedPoseDriverLookup` is null as well.

```cs
#pragma warning disable CS0618 // Type or member is obsolete
  // Find the controller lookup class in the hierarchy. Solvers that require access to the
  // left, right or gaze controllers will use the references stored in this class.
if (controllerLookup == null && trackedPoseDriverLookup == null)
{
    controllerLookup = ComponentCache<ControllerLookup>.FindFirstActiveInstance();
}
#pragma warning restore CS0618
```

**Review by AMollis — CHANGES_REQUESTED** [review]
Looks good some minor changes needed

File: org.mixedrealitytoolkit.core/Utilities/ControllerLookup.cs [hunk 6]
```diff
@@ -12,6 +13,7 @@ namespace MixedReality.Toolkit
     /// </summary>
     [DisallowMultipleComponent]
     [AddComponentMenu("MRTK/Core/Controller Lookup")]
+    [Obsolete]
```
[the remark this answers is not in the document]
Comment (ms-RistoRK) [reply org.mixedrealitytoolkit.core/Utilities/ControllerLookup.cs]: Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/749/commits/4ac34710955e2bcd471870a113bc95a4ed2c2ce5 , resolving.

File: org.mixedrealitytoolkit.core/Utilities/Experimental/TrackedPoseDriverLookup.cs
[the remark this answers is not in the document]
Comment (ms-RistoRK) [reply org.mixedrealitytoolkit.core/Utilities/Experimental/TrackedPoseDriverLookup.cs:1]: Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/749/commits/ae3356ded94a8b1e773729ffe66cd5f3610dec3e , resolving.

File: org.mixedrealitytoolkit.input/InteractionModes/FlatScreenModeDetector.cs [hunk 7]
```diff
@@ -18,14 +21,18 @@ internal class FlatScreenModeDetector : MonoBehaviour, IInteractionModeDetector
 
         public InteractionMode ModeOnDetection => flatScreenInteractionMode;
 
-        protected ControllerLookup controllerLookup = null;
+        [Obsolete]
```
[the remark this answers is not in the document]
Comment (ms-RistoRK) [reply org.mixedrealitytoolkit.input/InteractionModes/FlatScreenModeDetector.cs]: Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/749/commits/0ba2665ff76b4d8de4fd291de6c244ead64896d8 , resolving.

File: org.mixedrealitytoolkit.spatialmanipulation/Solvers/Solver.cs (hunk 2 above)
Comment (ms-RistoRK) [reply org.mixedrealitytoolkit.spatialmanipulation/Solvers/Solver.cs]: Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/749/commits/c16545c686519bafd6fb5ffd75360b780829a71f , resolving.

File: org.mixedrealitytoolkit.input/InteractionModes/FlatScreenModeDetector.cs [hunk 8]
```diff
@@ -34,7 +41,27 @@ protected void Awake()
         public bool IsModeDetected()
         {
             // Flat screen mode is only active if the Left and Right Hand Controllers aren't being tracked
-            return !controllerLookup.LeftHandController.currentControllerState.inputTrackingState.HasPositionAndRotation() && !controllerLookup.RightHandController.currentControllerState.inputTrackingState.HasPositionAndRotation();
+            if (controllerLookup != null) //Note: Will be removed for XRI3 migration completion
```
[the remark this answers is not in the document]
Comment (ms-RistoRK) [reply org.mixedrealitytoolkit.input/InteractionModes/FlatScreenModeDetector.cs:45]: Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/749/commits/2e7783a6407a5447c9ecefb943a266855b096514 , resolving.

File: org.mixedrealitytoolkit.spatialmanipulation/Solvers/HandConstraintPalmUp.cs (hunk 1 above)
Comment (ms-RistoRK) [reply org.mixedrealitytoolkit.spatialmanipulation/Solvers/HandConstraintPalmUp.cs:278]: Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/749/commits/2e7783a6407a5447c9ecefb943a266855b096514 , resolving.

File: org.mixedrealitytoolkit.spatialmanipulation/Solvers/Solver.cs (hunk 5 above)
Comment (ms-RistoRK) [reply org.mixedrealitytoolkit.spatialmanipulation/Solvers/Solver.cs:282]: Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/749/commits/2e7783a6407a5447c9ecefb943a266855b096514 and https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/749/commits/7d515a9520774e8b9be79bdd22ff5d7a59bf329c , resolving.

File: org.mixedrealitytoolkit.spatialmanipulation/Solvers/Solver.cs (hunk 4 above)
Comment (ms-RistoRK) [reply org.mixedrealitytoolkit.spatialmanipulation/Solvers/Solver.cs]: Right, that's a remnant of a first attempt that I forgot to remove, thx for catching that.  Fixed in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/749/commits/12cc9a84d29f25f7250686ede287c6d10e7726b9 , resolving.

File: org.mixedrealitytoolkit.input/InteractionModes/FlatScreenModeDetector.cs [hunk 9]
```diff
@@ -34,7 +41,29 @@ protected void Awake()
         public bool IsModeDetected()
         {
             // Flat screen mode is only active if the Left and Right Hand Controllers aren't being tracked
-            return !controllerLookup.LeftHandController.currentControllerState.inputTrackingState.HasPositionAndRotation() && !controllerLookup.RightHandController.currentControllerState.inputTrackingState.HasPositionAndRotation();
+            #pragma warning disable CS0618 // Type or member is obsolete
+            if (controllerLookup != null) //Note: Will be removed for XRI3 migration completion
```
Comment (AMollis) [inline org.mixedrealitytoolkit.input/InteractionModes/FlatScreenModeDetector.cs:45]: Remove `//Note: Will be removed for XRI3 migration completion`

**Review by AMollis — APPROVED** [review]
Approve, with change requests

File: org.mixedrealitytoolkit.input/InteractionModes/FlatScreenModeDetector.cs (hunk 9 above)
Comment (ms-RistoRK) [reply org.mixedrealitytoolkit.input/InteractionModes/FlatScreenModeDetector.cs]: Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/749/commits/f0cb2196739a51e664622c588a6771d0abd87940 , resolving.

File: org.mixedrealitytoolkit.spatialmanipulation/Solvers/HandConstraintPalmUp.cs [hunk 10]
```diff
@@ -274,21 +274,49 @@ private bool IsUserGazeMeetingThresholdRequirements(XRNode hand)
                 Ray? gazeRay = null;
                 bool usedEyeGaze = false;
 
-                if (ControllerLookup != null &&
-                    ControllerLookup.GazeController != null &&
-                    (ControllerLookup.GazeController.currentControllerState.inputTrackingState &
-                    (InputTrackingState.Position | InputTrackingState.Rotation)) > 0)
+                #pragma warning disable CS0618 // Type or member is obsolete
+                if (ControllerLookup != null) //Note: Will be removed for XRI3 migration completion
```
[the remark this answers is not in the document]
Comment (ms-RistoRK) [reply org.mixedrealitytoolkit.spatialmanipulation/Solvers/HandConstraintPalmUp.cs]: Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/749/commits/f0cb2196739a51e664622c588a6771d0abd87940 , resolving.

**Review by AMollis — APPROVED** [review]

**Pull request merged**